http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpRequest.java

```diff
@@ -134,18 +138,28 @@ default URI getUri() {
      * {@link SdkHttpFullRequest#builder()}.
      */
     interface Builder extends CopyableBuilder<Builder, SdkHttpRequest> {
+
         /**
-         * Convenience method to set the {@link #protocol()}, {@link #host()}, {@link #port()}, and
-         * {@link #encodedPath()} from a {@link URI} object.
+         * Convenience method to set the {@link #protocol()}, {@link #host()}, {@link #port()},
+         * {@link #encodedPath()} and optional via {@link #keepUriQueryParams()} query parameters from a {@link URI} object.
          *
-         * @param uri URI containing protocol, host, port and path.
+         * @param uri URI containing protocol, host, port, path and optional query parameters.
          * @return This builder for method chaining.
          */
         default Builder uri(URI uri) {
-            return this.protocol(uri.getScheme())
-                       .host(uri.getHost())
-                       .port(uri.getPort())
-                       .encodedPath(SdkHttpUtils.appendUri(uri.getRawPath(), encodedPath()));
+            SdkHttpRequest.Builder builder = this.protocol(uri.getScheme())
+                                                 .host(uri.getHost())
+                                                 .port(uri.getPort())
+                                                 .encodedPath(SdkHttpUtils.appendUri(uri.getRawPath(), encodedPath()));
+            if (uri.getQuery() != null) {
+                Pattern.compile("\\s*&\\s*")
+                       .splitAsStream(uri.getQuery().trim())
+                       .map(s -> s.split("=", 2))
+                       .collect(groupingBy(a -> a[0], mapping(a -> a[1], toList())))
+                       .forEach((paramKey, paramValues) -> paramValues
+                           .forEach(paramValue -> this.appendUriQueryParameter(paramKey, paramValue)));
```

Comment:
Think the query parameters need to be decoded as well

Reply:
@dagnir Thanks for the review. I've tried my best to update the upstream master but to have 2 commits now that I am unsure of how to properly squash into a single one (9d6fae0647 and 43a43be5a7)